`accept()`d socket needs to be set to `O_NONBLOCK`

jshearer opened the pull request:
I also copied the fail-handling logic from `sys/unix/net.rs` to actually close the socket on error, as described by the comment.

Closes #1450

Thomasdezeeuw commented:
I've commented on #1450, but I don't think this is the right solution as all OSs that do use the path should have the behaviour that inherits the non-blocking property of the listening socket. Android doesn't event follow the `accept(2)` path, but uses the `accept4(2)` with `SOCK_NONBLOCK` path.

Thomasdezeeuw commented on src/sys/unix/uds/listener.rs:
```diff
@@ -80,7 +80,13 @@ pub(crate) fn accept(listener: &net::UnixListener) -> io::Result<(UnixStream, So
-        syscall!(fcntl(socket, libc::F_SETFD, libc::FD_CLOEXEC)).map(|_| s)
+        syscall!(fcntl(socket, libc::F_SETFL, libc::O_NONBLOCK))
+            .and_then(|_| syscall!(fcntl(socket, libc::F_SETFD, libc::FD_CLOEXEC)))
+            .map_err(|e| {
```
This will close the socket twice as `s` of type `net::UnixStream` will also close the socket when dropped, so this can be removed.

Thomasdezeeuw commented on src/sys/unix/uds/listener.rs:
```diff
@@ -80,7 +80,13 @@ pub(crate) fn accept(listener: &net::UnixListener) -> io::Result<(UnixStream, So
         // error below.
         let s = unsafe { net::UnixStream::from_raw_fd(socket) };
-        syscall!(fcntl(socket, libc::F_SETFD, libc::FD_CLOEXEC)).map(|_| s)
+        syscall!(fcntl(socket, libc::F_SETFL, libc::O_NONBLOCK))
```
Can you limit this to x86 Android.

Thomasdezeeuw replied:
Can you also do this in `src/sys/unix/tcp.rs`, after `stream` is created on line 469. Also limited to x86 Android.

https://github.com/tokio-rs/mio/blob/527e843b0e11d585c0b20a8a3e4df966766f8bfa/src/sys/unix/tcp.rs#L461-L469

Thomasdezeeuw replied:
And finally can you add a comment referring to #1450.

jshearer commented:
I think I addressed your comments. My only concern is that my implementation might not be very idiomatic (`let mut` then mutating conditionally based on `cfg!`). Thoughts?

Thomasdezeeuw commented on src/sys/unix/tcp.rs:
```diff
@@ -465,7 +465,19 @@ pub fn accept(listener: &net::TcpListener) -> io::Result<(net::TcpStream, Socket
         .map(|socket| unsafe { net::TcpStream::from_raw_fd(socket) })
-        .and_then(|s| syscall!(fcntl(s.as_raw_fd(), libc::F_SETFD, libc::FD_CLOEXEC)).map(|_| s))
+        .and_then(|s| {
+            let mut f = syscall!(fcntl(s.as_raw_fd(), libc::F_SETFD, libc::FD_CLOEXEC));
```
You can use `syscall!(...)?` here and ignore the result (now stored in `f`).

jshearer replied:
Do you mean like this? Because this actually doesn't seem to work 🤔 
```rust
let stream = {
        syscall!(accept(
            listener.as_raw_fd(),
            addr.as_mut_ptr() as *mut _,
            &mut length
        ))
        .map(|socket| unsafe { net::TcpStream::from_raw_fd(socket) })
        .and_then(|s| {
            syscall!(fcntl(s.as_raw_fd(), libc::F_SETFD, libc::FD_CLOEXEC))?;
    
            // See https://github.com/tokio-rs/mio/issues/1450
            #[cfg(all(arget_arch = "x86", target_os = "android"))]
            syscall!(fcntl(s.as_raw_fd(), libc::F_SETFL, libc::O_NONBLOCK))?;
            
            Ok(s)
        })
    }?;
```

jshearer replied:
`arget_arch` d'oh

It works fine 🙄

Thomasdezeeuw commented on src/sys/unix/tcp.rs:
```diff
@@ -465,7 +465,19 @@ pub fn accept(listener: &net::TcpListener) -> io::Result<(net::TcpStream, Socket
+            let mut f = syscall!(fcntl(s.as_raw_fd(), libc::F_SETFD, libc::FD_CLOEXEC));
+    
+            // See https://github.com/tokio-rs/mio/issues/1450
+            if cfg!(all(
```
No need to use `if`, the following should work.

```
#[cfg(all(arget_arch = "x86", target_os = "android"))
syscall!(fcntl(s.as_raw_fd(), libc::F_SETFL, libc::O_NONBLOCK))?;
```

Thomasdezeeuw commented on src/sys/unix/tcp.rs:
```diff
@@ -465,7 +465,19 @@ pub fn accept(listener: &net::TcpListener) -> io::Result<(net::TcpStream, Socket
+                f = f.and_then(|_| syscall!(fcntl(s.as_raw_fd(), libc::F_SETFL, libc::O_NONBLOCK)));
+            }
+            
+            f.map(|_| s)
```
This can be `Ok(s)`.

jshearer commented:
After fixing my typo, tests pass on Android with your suggestions :)

Thomasdezeeuw commented:
Thanks @jshearer.
